src/meili/map_matcher.cc

```diff
@@ -202,88 +201,82 @@ MatchResult FindMatchResult(const MapMatcher& mapmatcher,
-    //                      |
-    //   edge (prev_state)  |   edge (next state)
-    // -------------------->N1------------------>N2
-    //                      |
-    //                      |
-    //                      |
-
-    bool at_intersection = edge.percent_along == 0.0 || edge.percent_along == 1.0;
-    if (!at_intersection && (edge.id == prev_state_edgeid || edge.id == next_state_edgeid)) {
+
+    // if it matches either end of the path coming into this state or the beginning of the
+    // path leaving this state, then we are good to go and have found the match
+    if (edge.id == prev_edge || edge.id == next_edge) {
       return {edge.projected,     std::sqrt(edge.distance), edge.id,
               edge.percent_along, measurement.epoch_time(), stateid};
     }
-    // last matched location is on intersection:
-    // If there is a valid edge in previous state. For leg builder to work properly in
-    // intersections (where edge candidates may not be correct), we snap the matched
-    // location to the previous edge's end and set the percentage along to 1.0
-    // p.s: this also handles any middle point that are matched on intersections
-    else if (prev_state_edgeid.Is_Valid() && at_intersection &&
-             (prev_end == curr_edge_start || prev_end == curr_edge_end)) {
-      // curr rout could be one way with edge ending at the same node with previous edge
-      return {edge.projected,           std::sqrt(edge.distance),
-              prev_state_edgeid,        1.0,
-              measurement.epoch_time(), stateid};
+
+    // the only matches we can make where the ids arent the same are at intersections
+    if (edge.percent_along > 0.f && edge.percent_along < 1.f) {
+      continue;
     }
-    // first matched location is on intersection:
-    // If there is a valid edge in next state. For leg builder to work properly in
-    // intersections (where edge candidates may not be correct), we snap the matched
-    // location to the next edge's start and set the percentage along to 0.0
-    else if (next_state_edgeid.Is_Valid() && at_intersection &&
-             (next_start == curr_edge_end || next_start == curr_edge_start)) {
-      return {edge.projected,           std::sqrt(edge.distance),
-              next_state_edgeid,        0.0,
-              measurement.epoch_time(), stateid};
+
+    // we are at an intersection so we need the node from the candidate where the route would be
+    auto candidate_nodes = graph_reader.GetDirectedEdgeNodes(edge.id, tile);
+    const auto& candidate_node =
+        edge.percent_along == 0.f ? candidate_nodes.first : candidate_nodes.second;
+
+    // try previous and next (first one and discontinuities wont have previous)
+    for (const auto& path_edge : {prev_edge, next_edge}) {
+      // is the candidate node the begin node of this edge
+      auto nodes = graph_reader.GetDirectedEdgeNodes(path_edge, tile);
+      if (nodes.first == candidate_node) {
```

Comment:
this brings up another question. could this lead to false positives. the edgelabel itself doesnt tell us whether it was for a node candidate. so its possible some of the candidates were node candidates but the route that was selected was not, but does share a node with the node candidate edge.

i feel pretty certain that this is a put fall here. it seems to me, we need to first try to find a non-intersection match result and then only look at intersection ones last.. so i can refacter the code to do that.

Reply:
ok nevermind i realized that the last or first label that routing.cc will use in the case it is at a node will be the right one with respect to whether it begins or ends a route and so i changed it to only check for previous edge ending at node and next edge beginning at node.